components/esp32_ble_presense/esp32_ble_presense.cpp

```diff
@@ -0,0 +1,139 @@
+////////////////////////////////////////////////////////////////////////
+// SPDX-License-Identifier: AGPL-3.0-only
+// (C) Copyright 2022-2024 - formatBCE
+
+#include "esp32_ble_presense.h"
+
+#include "esphome/core/log.h"
+
+#include <Arduino.h>
+#include <NimBLEDevice.h>
+
+#define bleScanInterval 0x80 // Used to determine antenna sharing between Bluetooth and WiFi. Do not modify unless you are confident you know what you're doing
+#define bleScanWindow 0x40 // Used to determine antenna sharing between Bluetooth and WiFi. Do not modify unless you are confident you know what you're doing
+
+using namespace esphome;
+
+namespace ESP32_BLE_Presense {
+
+NimBLEScan* pBLEScan;
+
+class BleAdvertisedDeviceCallbacks: public BLEAdvertisedDeviceCallbacks {
+
+    ESP32_BLE_Presense& parent_;
+
+public:
+    BleAdvertisedDeviceCallbacks(ESP32_BLE_Presense& parent) : parent_(parent) {}
+
+    void onResult(NimBLEAdvertisedDevice* device) {
+        if (!device)
+            return;
+
+        parent_.reportDevice(device->getAddress().toString(),
+                             device->getRSSI(),
+                             device->getManufacturerData());
+    }
+};
+
+// Copies unneccessarily
+static std::string capitalizeString(std::string s) {
+    std::transform(s.begin(), s.end(), s.begin(),
+                   [](unsigned char c){ return std::toupper(c); });
+    return s;
+}
+
+static unsigned long getTime() {
+    time_t now;
+    struct tm timeinfo;
+    if (!getLocalTime(&timeinfo)) {
+        return(0);
+    }
+    ::time(&now);
+    return now;
+}
+
+ESP32_BLE_Presense::ESP32_BLE_Presense()
+:  esphome::PollingComponent(5000) {
```

Comment:
This is the only change that I think could be of interest.  Instead of initiating the scan from a separate task this is run periodically from one of the main task.  However since NimBLE seems to run in its own task I don't expect this to cause any problems.

Reply:
Hopefully. Because running it in separate task on separate core was the main difference between this and inbuilt ESPHome BLE scanner - and that thing is terrible. :)

Reply:
good to know - spawning a new task to initiate the polling shouldn't be hard if needed.

Reply:
It's rather the core, as far as I understand. But I think everything will be alright. NimBLE handles that, most probably.